### ISSUE TITLE
Duplicate classes warnings at runtime when multiple targets are contained in the same project

### ISSUE THREAD

**Issue by combinatorial** [CocoaPods/CocoaPods#7155]
* [x] I've read and understood the [*CONTRIBUTING guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report

## What did you do?

Run the app

## What did you expect to happen?

The app to run without console warnings about multiple classes.

## What happened instead?

Console warnings of the type 

`class AFAutoPurgingImageCache is implemented in both /Users/david/Library/Developer/Xcode/DerivedData/PodsDynamicLibTest-gidixodenwoqttcqkxresgoadfkg/Build/Products/Debug-iphonesimulator/Framework.framework/Framework (0x102eaac58) and /Users/david/Library/Developer/CoreSimulator/Devices/8ED2F0F1-37B0-4BE2-8427-B4D70DEE1A2D/data/Containers/Bundle/Application/0868D1B5-9454-4F07-8CDE-5218F53901A8/App.app/App (0x101f9f050). One of the two will be used. Which one is undefined.`

### Stack

```
   CocoaPods : 1.3.1
        Ruby : ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]
    RubyGems : 2.6.11
        Host : Mac OS X 10.12.6 (16G29)
       Xcode : 9.1 (9B46)
         Git : git version 2.10.1
Ruby lib dir : /usr/local/Cellar/ruby/2.4.1_1/lib
Repositories : helpshift - https://github.com/helpshift/CocoaPods.git @ f32e7f82986f8df58f5cbb30421a8548eb222b21
               master - https://github.com/CocoaPods/Specs.git @ bc25b40465df84e6645d99b0c606ac7b102e2eda
               tophatch - https://github.com/tophatch/SpecsRepo.git @ 778d9226d877dd3ffa3ccae9d8baa7c8cb1db278
```

### Installation Source

```
Executable Path: /usr/local/bin/pod
```

### Plugins

```
cocoapods-clean        : 0.0.1
cocoapods-deintegrate  : 1.0.1
cocoapods-dependencies : 1.0.0
cocoapods-plugins      : 1.0.0
cocoapods-search       : 1.0.0
cocoapods-stats        : 1.0.0
cocoapods-trunk        : 1.2.0
cocoapods-try          : 1.1.0
```

### Podfile

```ruby
source 'https://github.com/CocoaPods/Specs.git'
workspace 'PodsDynamicLibTest.xcworkspace'

target 'Framework' do
    platform :ios, '11.0'
    project 'App/App.xcodeproj'
    pod 'AFNetworking', '~> 3.0'
end

target 'App' do
    platform :ios, '11.0'
    project 'App/App.xcodeproj'
    pod 'DateTools'
end
```

## Project that demonstrates the issue

https://github.com/tophatch/CocoapodsDuplicateSymbolTestProject

The issue occurs when the two targets are embedded in the same parent project. If the targets are in different projects then it does not occur.

This was not an issue in cocoapods 1.2, it was introduced in 1.3.

If you look at https://github.com/tophatch/CocoapodsDuplicateSymbolTestProject/blob/master/Pods/Target%20Support%20Files/Pods-App/Pods-App.debug.xcconfig
you will see that OTHER_LDFLAGS includes `-l"AFNetworking"` I think it should not be included.

**Comment by combinatorial** [comment]
This was actually introduced in 1.2.1. Perhaps with #6481?

**Comment by paulb777** [comment]
It does seem wrong for the App to link in `-l"AFNetworking"`, but it also seems a bit strange to want to link a dynamic framework but not include `use_frameworks!` in the Podfile.  Adding `use_frameworks!` will also eliminate the warnings.

**Comment by combinatorial** [comment]
My understanding is that `use_frameworks!` causes dynamic frameworks to be used for each pod that are dependencies, is that correct? If so, then this has a performance impact on start-up for doing this. The desired behavior is to statically link all of the pods into the single dynamic framework.

**Comment by paulb777** [comment]
Yes, `use_frameworks!` causes a dynamic framework for each dependency pod.

What's the reason for one dynamic framework versus linking everything statically?

**Comment by combinatorial** [comment]
In our case we have multiple app extensions that share a common core with the main app.

**Comment by paulb777** [comment]
Understood, thanks.

It does look related to #6481. 

The `-l"AFNetworking"` gets removed and the warnings go away with the following coarse-grained change:
```
$ git diff
diff --git a/lib/cocoapods/generator/xcconfig/xcconfig_helper.rb b/lib/cocoapods/generator/xcconfig/xcconfig_helper.rb
index 22fc4d20..84ea4439 100644
--- a/lib/cocoapods/generator/xcconfig/xcconfig_helper.rb
+++ b/lib/cocoapods/generator/xcconfig/xcconfig_helper.rb
@@ -461,8 +461,8 @@ module Pod
           other_ld_flags = pod_targets.select(&:should_build?).map do |pod_target|
             if pod_target.requires_frameworks?
               %(-framework "#{pod_target.product_basename}")
-            elsif XCConfigHelper.links_dependency?(aggregate_target, pod_target)
-              %(-l "#{pod_target.product_basename}")
+        #    elsif XCConfigHelper.links_dependency?(aggregate_target, pod_target)
+        #      %(-l "#{pod_target.product_basename}")
             end
           end
           xcconfig.merge!('OTHER_LDFLAGS' => other_ld_flags.compact.join(' '))
```
Any thoughts @dnkoutso?

**Comment by dnkoutso** [comment]
This was mostly an issue fix in which Cocoapods links the same dependency for both test bundles and app targets.

Theoretically dynamic frameworks can link dependencies and the app target should not link them but the cocoapods model since the beginning is that all dependencies are linked to the app (host) target instead.

There is a chance this can be fixed by a smaller change but I think we need a bigger plan in which dynamic frameworks actually link their dependencies.

**Comment by dnkoutso** [comment]
If a dependency is linked and added to the app target then it should not be added to the test target. That was the initial fix with `links_dependency?` as i recalled correctly and I think it was part of 1.2.0 release.

**Comment by combinatorial** [comment]
Is issue something that will likely be fixed? Is there anything I can do to help?

**Comment by dnkoutso** [comment]
Its fairly complicated issue, I think the entire system with embedding targets is fairly broken at the moment, everything is assumed to be linking to the top level target.

I think a general overhaul of this linking mechanism is required. 

The other issue I had mentioned earlier is double linking dependencies on test bundles and app targets which 1.3.x fixes. We should not be linking dependencies to a test bundle if the app already brings that dependencies otherwise you get multiple warnings around class loading.

**Comment by marchy** [comment]
The problem is that to use app extensions projects need to be split up into multiple frameworks/layers (ie: at least a core one that can be consumed from both the main app and the respective app extensions), however at runtime certain 3rd party libraries will fail (for example GoogleMaps) because they try to access static variables and the static values seem not to be shared across framework. In other words there will be different versions of the class definition at runtime. This is a huge problem – the impact being that shared code that uses any sort of class/static variable cannot inter-operate across the app frameworks!

**Comment by hongrich** [comment]
Posting the workaround that I'm currently using in case it's helpful to anyone. This workaround uses the `pre_install` hook to remove duplicate dependencies. The issue occurs in the analyzer which runs before the `pre_install` hook. So you only have to change one thing if you make the change there. Using the example from https://github.com/CocoaPods/CocoaPods/issues/7155#issue-267398054, add the following to your `Podfile`.

```ruby
pre_install do |installer|
  embedded_target = installer.aggregate_targets.find { |aggregate_target| aggregate_target.name == 'Pods-Framework' }
  host_target = installer.aggregate_targets.find { |aggregate_target| aggregate_target.name == 'Pods-App' }
  host_target.pod_targets = host_target.pod_targets - embedded_target.pod_targets
end
```

From what I can tell, this issue was first introduced in https://github.com/CocoaPods/CocoaPods/pull/6575 and first released as v1.2.1.

Please let me know if you notice any potential issues with my workaround.

Related issues:
* https://github.com/CocoaPods/CocoaPods/issues/6918
* https://github.com/CocoaPods/CocoaPods/issues/7126

**Comment by combinatorial** [comment]
This problem still exists in cocoapods-1.5.3

**Comment by yoon-boom** [comment]
@combinatorial have you found any solution to this?

**Comment by combinatorial** [comment]
No I have not.

**Comment by yoon-boom** [comment]
@combinatorial Looks like long time Google search, I found a solution on my case similar to this. This might not apply to your case.

Previously, I also receives multiple target error messages (`One of the two will be used. Which one is undefined.`) and crashes from test. I had two targets (**App** and **App Test**), used pod version `1.5.2` and some said remove `OTHER_LDFLAGS` from `Pods-[AppName].debug.xcconfig` which worked for me. ([HERE](https://stackoverflow.com/questions/46932341/class-is-implemented-in-both-one-of-the-two-will-be-used-which-one-is-undefine?rq=1)) But I believe it doesn't seem right things to do, especially each time updating pods. So here is what I tried. 

1) Go to project setting select target;In my case TestTarget Unit Test.
2) Build Settings->Other Linker Flags;It had listed all the frameworks from Cocoapods (which I believe already ported on Host Project Target)
3) Remove the duplicate lists, from release candidates (Debug and/or Release)
4) Try to update the pod by `pod update`
5) It will still include `OTHER_LDFLAGS`, but no more complain about duplicate warnings on runtime

**Comment by mehdok** [comment]
is There any concrete workaround to Fix this problem?

**Comment by bestwnh** [comment]
@mehdok https://stackoverflow.com/a/52773670/2446684 Hope this can help you.

**Comment by paulb777** [comment]
@pnemere It is dangerous to build with the duplicate warnings.  I recommend using one of the workarounds discussed above until there is a more general CocoaPods fix.

**Comment by amorde** [comment]
@paulb777  should this be re-opened?

**Comment by paulb777** [comment]
@amorde Yes I confirmed the original test case still fails on current master.

**Comment by Yetispapa** [comment]
Any update on this. Getting the same errors for Fabric, BrainTree and Flurry

**Comment by ileitch** [comment]
```ruby
PROJECT_ROOT_DIR = File.dirname(File.expand_path(__FILE__))
PODS_DIR = File.join(PROJECT_ROOT_DIR, 'Pods')
PODS_TARGET_SUPPORT_FILES_DIR = File.join(PODS_DIR, 'Target Support Files')

post_install do |installer|
  remove_static_framework_duplicate_linkage({
    'SharedFramework' => ['Fabric', 'Crashlytics', 'GoogleMaps', 'GoogleSignIn', 'GooglePlaces']
  })
end

# CocoaPods provides the abstract_target mechanism for sharing dependencies between distinct targets.
# However, due to the complexity of our project and use of shared frameworks, we cannot simply bundle everything under
# a single abstract_target. Using a pod in a shared framework target and an app target will cause CocoaPods to generate
# a build configuration that links the pod's frameworks with both targets. This is not an issue with dynamic frameworks,
# as the linker is smart enough to avoid duplicate linkage at runtime. Yet for static frameworks the linkage happens at
# build time, thus when the shared framework target and app target are combined to form an executable, the static
# framework will reside within multiple distinct address spaces. The end result is duplicated symbols, and global
# variables that are confined to each target's address space, i.e not truly global within the app's address space.
#
# Previously we avoided this by linking the static framework with a single target using an abstract_target, and then
# provided a shim to expose their interfaces to other targets. The new approach implemented here removes the need for
# shim by modifying the build configuration generated by CocoaPods to restrict linkage to a single target.
def remove_static_framework_duplicate_linkage(static_framework_pods)
  puts "Removing duplicate linkage of static frameworks"

  Dir.glob(File.join(PODS_TARGET_SUPPORT_FILES_DIR, "Pods-*")).each do |path|
    pod_target = path.split('-', -1).last

    static_framework_pods.each do |target, pods|
      next if pod_target == target
      frameworks = pods.map { |pod| identify_frameworks(pod) }.flatten

      Dir.glob(File.join(path, "*.xcconfig")).each do |xcconfig|
        lines = File.readlines(xcconfig)

        if other_ldflags_index = lines.find_index { |l| l.start_with?('OTHER_LDFLAGS') }
          other_ldflags = lines[other_ldflags_index]

          frameworks.each do |framework|
            other_ldflags.gsub!("-framework \"#{framework}\"", '')
          end

          File.open(xcconfig, 'w') do |fd|
            fd.write(lines.join)
          end
        end
      end
    end
  end
end
```

**Comment by Yetispapa** [comment]
Where should I paste in this snippet? In my podfile?

**Comment by ileitch** [comment]
@Yetispapa yes, you also need to change the call to `remove_static_framework_duplicate_linkage` to match your project configuration.

**Comment by Yetispapa** [comment]
I tried the code but I got an error on the identify_framework line. I just put the snippet below my framework list. Is that right? Can you give me an simple example how the pod file should look like. Thanks 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by bestwnh** [comment]
@Yetispapa Maybe you could try my solution. I used it in commercial projects for long time.
I write a blog article to make it clear:
https://medium.com/@GalvinLi/tinysolution-resolve-the-restrictions-of-enum-d854ba49886d
And a demo project for reproduce the issue and try the fix: https://github.com/bestwnh/TinySolution

**Comment by astrokin** [comment]
@Yetispapa I have the same. @ileitch could you please add `identify_frameworks` function?

**Comment by ileitch** [comment]
@Yetispapa @astrokin Oops, updated with ``identify_frameworks`.

**Comment by jakunico** [comment]
We're currently phasing this issue, used the [solution](https://github.com/CocoaPods/CocoaPods/issues/7155#issuecomment-461395735) from @ileitch which worked good.

Is there any official workaround/fix planned in future Cocoapods releases?

**Comment by plivesey** [comment]
@jakunico  The solution from @ileitch doesn't work for me. I get:
```
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_GIDSignIn", referenced from:
      objc-class-ref in GoogleLogin.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I have a non-cocoapods framework in an app. I have included the pod in both (GoogleSignIn). Without the code it runs with a duplicate framework. With the extra Podfile code, it won't compile.

```$ pod --version
1.7.0.rc.1```

**Comment by cgmaier** [comment]
I've seen a fair number of workarounds for this on the app-side (adding stuff to Podfile) around the web, but does anyone know of any way to configure the dynamic framework such that devs who install via Cocoapods won't see these warnings?

**Comment by smoothdvd** [comment]
Does latest version fix this issue?

**Comment by a83988029** [comment]
This kind of warnings will make singleton not working.
We made the subprojects static libs to avoid this problem.

**Comment by AKoulabukhov** [comment]
But this issue isn’t resolved yet

**Comment by soxjke** [comment]
Does anybody of the contributors know, it this is on track to be included in 1.9.0 ?

**Comment by amorde** [comment]
It's not - if this is an issue for you I'd encourage you to contribute 👍 

1.9.0 beta is wrapping up soon so I don't think this would make it in to that release.

**Comment by nebitrams** [comment]
I am using Xcode 11.3 with cocoapods 1.9.1 (latest).  I am still facing this issue when I do Product -> Archive.  This error only happen during archiving and it works fine during Running,Testing.

Multiple targets match implicit dependency for linker flags '-framework Parse'. Consider adding an explicit dependency on the intended target to resolve this ambiguity. (in target 'App1' from project 'App1')

### Work around
I found a workaround for this but this is cumbersome because I need to run pod install to switch compilation for Testing vs Archiving.

### My original Podfile

```
target 'App1' do
  ...
  pod 'Parse', '1.17.3'
end

target 'App1Test' do
  ...
  pod 'Parse', '1.17.3'
end
```

### Workaround Podfile
I used the work around suggested by Huang Diaobo in [stack overflow](https://stackoverflow.com/questions/58407257/xcode-warning-multiple-targets-match-implicit-dependency-for-product)

```
target 'App1' do
  ...
  pod 'Parse', '1.17.3'
end

schema = ENV['schema']
if schema == 'UnitTest'
  target 'App1Test' do
    pod 'Parse', '1.17.3'
  end
end
```

### How to switch installation?
With the above, I need to switch the pod install whenever I build for test vs build for archive.
1) For UnitTest: export schema='UnitTest' && pod install
2) For Archiving: unset schema && pod install

**Comment by haofanning** [comment]
The issue still exist

**Comment by filip982** [comment]
This issue still exists and above solutions are not helping. Could we do something about it? It is a big problem when we want to modularize the project into frameworks. Thanks for all hard work

**Comment by rajdhakate** [comment]
I'm also facing the same issue.

**System**

`MacOS Catalina 10.15.3`

`Xcode: 11.4.1`

`Cocoapods : 1.9.1`

**Podfile defination**

platform :ios, '10.0'

def shared_pods
pod 'Firebase/Messaging'
end

target 'MainTarget' do
  use_frameworks!
  pod 'Moya', '~> 14.0'
  shared_pods
  pod 'Firebase/Auth'
  pod 'Firebase/Analytics'
  pod 'Firebase/Database'
  pod 'Firebase/Crashlytics'
  pod 'GoogleSignIn'
  pod 'FBSDKLoginKit'
end

target 'NotificationService' do
  use_frameworks!
  shared_pods
end

**Warning**

I'm using multiple Firebase pods in MainTarget. And in an app extension target using pod 'Firebase/Messaging'. Getting warning : `"Multiple target match implicit dependency for linker flag -framework GoogleUtilities"` for both targets. However running in simulator works fine.

**Error**

While archiving the project I'm getting error `Multiple command produce ...`. Can not archive the project

**Comment by Warpzit** [comment]
Closed but the bug is still present in CocoaPods 1.9.3.

**Comment by StasanTelnov** [comment]
Hello

I have a same problem, but only with `GoogleMaps` pod. I downgrade CocoaPods to 1.9.2 version, but problem actual. 

```
My pod file:
platform :ios, '10.0'

use_frameworks!
inhibit_all_warnings!
source 'https://github.com/CocoaPods/Specs.git'


def shared_pods
#     pod 'Kingfisher'
#     pod 'GoogleMaps'
    pod 'Fabric'
    pod 'Crashlytics'
    pod 'SVProgressHUD'
    pod 'IQKeyboardManagerSwift'
    pod 'VoxImplantSDK'
    pod 'Firebase/Core'
    pod 'AppsFlyerFramework'
    pod 'Moya'
#     pod 'ObjectMapper'
#     pod 'SnapKit'
end


abstract_target 'defaults' do

# main
	target 'MainTarget' do
		shared_pods
	end
	target 'MainTarget-dev' do
		shared_pods
		pod 'EnvironmentSwitcher'
	end


# modules	
	target 'LiteOverlayView' do
		pod 'SnapKit'
	end
	
	target 'LiteRoundedButton' do
		pod 'SnapKit'
	end
	
	target 'LiteExtensions' do
		pod 'Kingfisher'
		pod 'GoogleMaps'
		pod 'ObjectMapper'
	end
end
```

In pod file `MainTarget` and `MainTarget-dev` is usual targets, other `Lite*` targets - inner is frameworks

And on run in console I see:
```
objc[4688]: Class GMSMapsClearcutClient is implemented in both /private/var/containers/Bundle/Application/2F37632C-EE38-448F-94F6-B8FF29EED7C5/lite.app/Frameworks/LiteExtensions.framework/LiteExtensions (0x10485c5d8) and /private/var/containers/Bundle/Application/2F37632C-EE38-448F-94F6-B8FF29EED7C5/lite.app/lite (0x101b51a28). One of the two will be used. Which one is undefined.
objc[4688]: Class GMSAPIClientParameters is implemented in both /private/var/containers/Bundle/Application/2F37632C-EE38-448F-94F6-B8FF29EED7C5/lite.app/Frameworks/LiteExtensions.framework/LiteExtensions (0x10485c628) and /private/var/containers/Bundle/Application/2F37632C-EE38-448F-94F6-B8FF29EED7C5/lite.app/lite (0x101b51a78). One of the two will be used. Which one is undefined.
....
```
And it is only with `GoogleMaps` pod. Others (`SnapKit` and `ObjectMapper`) dont have classes, which `implemented in both`.
After display maps on view controller, I have runtime crash.


I see and try remove `GoogleMaps` manually from Build settings from `Other Linker Flags` and `Other C Flags`, but is dont works.

How I can fix this duplication or maybe downgrade to CocoaPods version?

**Comment by fkirc** [comment]
This is still incredibly broken with Firebase, but the following workaround from @TadeasKriz works for me: https://github.com/CocoaPods/CocoaPods/issues/7126#issuecomment-399395611

**Comment by bentonquest** [comment]
Hi!
Has anyone been able to fix this without the use of a `post_install` script?
I am trying to distribute a dynamic framework without the need for the user to add a script.
Any way to solve this on the podspec?
